Group G: Added bitset implementation

TsvetelinKostadinv opened the pull request:
Abstracted up the ContainerType so it can be easily swapped(added necessary static_assert for ease of use)
Added checks-throws/asserts in named methods/operators
Added the remaining methods of the big 4 - the d-tor, copy c-tor and operator= (might add move and swap)
Added flip operation on given index - standard bitset operation
Added explanatory comment in the main inside the for loop with alternatives
Added second bitset so it is easier to step through and examine what is happening
Added third bitset with zero size to illustrate throwing behavior with access outside bounds
Added convenience operators to the proxy
Added check-throw in the proxy constructor
Made use of `constexpr` so some values can be computed at compile-time

TsvetelinKostadinv committed:
Added bitset implementation
Abstracted up the ContainerType so it can be easily swapped(added necessary static_assert for ease of use)
Added checks-throws/asserts in named methods/operators
Added check-throw if the count of requested bits is 0 - no sense in a bitset with 0 bits
Added the remaining methods of the big 4 - the d-tor, copy c-tor and operator= (might add move and swap)
Added flip operation on given index - standard bitset operation
Added explanatory comment in the main inside the for loop with alternatives
Added second bitset so it is easier to step through and examine what is happening
Added conveniece operators to the proxy
Added check-throw in the proxy constructor
Made use of constexpr so some values can be computed at compile-time

TsvetelinKostadinv committed:
Removed const on the bitCount so we can operator= bitsets of different arity
Added cast in getMaskForIdx so we don't get warnings when we use anything other than uint32_t
Removed misleading comment
Added explanatory comment for the missing error handling logic behind the copying of the ContainerType - it is asserted to be integral and thus operator= cannot throw

TsvetelinKostadinv committed:
Added unsigned assert so we don't mess up with shifts

poseidon4o commented on group-G/source/proxy/proxy.cpp:
```diff
@@ -1,93 +1,251 @@
+		if (bitCount == 0)
+		{
+			delete[] data;
+			throw std::exception("A static bitset cannot be initialized with a count of zero(doesn't make sense)!");
```
Май е по-добре да оставиш възможнността да са празни? Това работи за заделянето на памет.

poseidon4o commented on group-G/source/proxy/proxy.cpp:
```diff
@@ -1,93 +1,251 @@
+		: data(new ContainerType[getCountOfContainerType(other.bitCount)])
+		, bitCount(other.bitCount)
+	{
+		for (size_t i = 0; i < getCountOfContainerType(other.bitCount); ++i)
+		{
+			// no concerns as Container type is integral => trivially (operator=)-able
+			data[i] = other.data[i];
 		}
```
Ще бъде по ефикасно да е memcpy.

poseidon4o commented on group-G/source/proxy/proxy.cpp:
```diff
@@ -1,93 +1,251 @@
+		delete[] data;
+		data = newData;
+		bitCount = other.bitCount;
+	}
```
Липсва return

poseidon4o commented on group-G/source/proxy/proxy.cpp:
```diff
@@ -1,93 +1,251 @@
 			, index(index)
-		{}
+		{
+			if (index < 0 || bitset->size() <= index)
+			{
+				throw std::out_of_range("The proxy must point to an element in the bitset!");
+			}
```
Май няма нужда от това, proxy винаги минава през метод който има тази проверка?

TsvetelinKostadinv replied:
При работа със самия битсет мисля, че ще е по-добре още при самото създаване на проксито да се сигнализира, че е извън битсета - колкото по-рано, толкова по-добре. Така може да си се поддържа вътрешен инвариант за проксито и да използва например методи без проверките?

poseidon4o replied:
Мен ми се иска обекти от proxy-то да не могат да живеят повече от израза в който са използвани, но не знам дали може да стане. Остави го така аз трябва да експериментирам с кода.

TsvetelinKostadinv replied:
За да живеят до края на израза(до `;`) до колкото знам може да се използва връщане на `proxy&`, което ще се поддържа живо до края на израза.

poseidon4o commented on group-G/source/proxy/proxy.cpp:
```diff
@@ -1,93 +1,251 @@
+			return *this;
+		}
+
+		void flip()
+		{
+			bitset->flip(index);
+		}
```
Това е окей, но на мен би ми било странно защото proxy-то се стреми да имитира 1 елемент (бит).

poseidon4o commented on group-G/source/proxy/proxy.cpp:
```diff
@@ -1,93 +1,251 @@
+		assert(0 <= index && index < bitCount && "The index has to be inside the bitfield");
 		return get(index);
 	}
+private:
+	static size_t getCountOfContainerType(int bitCount)
+	{
+		return (bitCount + containerTypeBitsSize - bitCount % containerTypeBytesSize) / containerTypeBitsSize;
+	}
+
+	constexpr static ContainerType getMaskForIdx(size_t bitIndex)
+	{
+		assert(bitIndex <= containerTypeBitsSize && "Cannot mask outside the bits of the container type!");
+		return ContainerType(1) << bitIndex;
+	}
+
+	struct IndexingPair
+	{
+		const size_t containerTypeIndex;
+		const size_t bitIndex;
+
+		ContainerType getMask() const { return getMaskForIdx(bitIndex); }
+		ContainerType getMaskInverted() const { return ~getMask(); }
+	};
+
+	static IndexingPair getIndexingPairForBitIndex(int idx)
+	{
+		return IndexingPair{ idx / containerTypeBitsSize, idx % containerTypeBitsSize };
+	}
+
```
Добави по кратък коментар какво правят новите методи и структурата.

TsvetelinKostadinv committed:
Removed the check-throw for zero-sized bitsets
Added example of throwing with access outside bounds

TsvetelinKostadinv committed:
Moved getting implementation to operator[]

TsvetelinKostadinv committed:
Changed brackets of the proxy constructor

TsvetelinKostadinv committed:
Changed for-loop to memcpy for efficiency

TsvetelinKostadinv committed:
Added missing return in operator=

TsvetelinKostadinv committed:
Added explanatory comments to the helper functions
Changed assert in getting the mask - 1 << 32 does not work as expected
Added constexpr's where applicable
Renamed function argument to be more descriptive

TsvetelinKostadinv committed:
Added new line at end of file
